gtsfm/two_view_estimator.py

```diff
@@ -114,19 +162,88 @@ def create_computation_graph(
         else:
             pose_error_graphs = (None, None)
-            corr_error_graph = None
-
-        return (
-            i2Ri1_graph,
-            i2Ui1_graph,
-            v_corr_idxs_graph,
-            pose_error_graphs[0],
-            pose_error_graphs[1],
-            corr_error_graph,
+            number_correct, inlier_ratio = None, None
+
+        R_error_deg, U_error_deg = pose_error_graphs[0], pose_error_graphs[1]
+
+        two_view_report_graph = dask.delayed(generate_two_view_report)(
+            inlier_ratio_est_model,
+            R_error_deg,
+            U_error_deg,
+            number_correct,
+            inlier_ratio,
+            v_corr_idxs_graph
         )
 
+        result = dask.delayed(self.check_for_degeneracy)(
+            two_view_report_graph, i2Ri1_graph, i2Ui1_graph, v_corr_idxs_graph
+        )
+        i2Ri1_graph, i2Ui1_graph, v_corr_idxs_graph, two_view_report_graph = result[0], result[1], result[2], result[3]
+
+        return (i2Ri1_graph, i2Ui1_graph, v_corr_idxs_graph, two_view_report_graph)
+
+    def check_for_degeneracy(
+        self,
+        two_view_report: TwoViewEstimationReport,
+        i2Ri1: Optional[Rot3],
+        i2Ui1: Optional[Unit3],
+        v_corr_idxs: np.ndarray,
+    ) -> Tuple[Optional[Rot3], Optional[Unit3], np.ndarray]:
+        """ """
+        insufficient_inliers = two_view_report.num_inliers_est_model < self._min_num_inliers_acceptance
+
+        H_EF_inlier_ratio = two_view_report.num_H_inliers / (two_view_report.num_inliers_est_model + EPSILON)
+        is_planar_or_panoramic = H_EF_inlier_ratio > MAX_H_INLIER_RATIO
+
+        # TODO: technically this should almost always be non-zero, just need to move up to earlier
+        valid_model = two_view_report.num_inliers_est_model > 0
+        if valid_model:
+            logger.info("H_EF_inlier_ratio: %.2f", H_EF_inlier_ratio)
+
+        if (valid_model and is_planar_or_panoramic) or (valid_model and insufficient_inliers):
+
+            if is_planar_or_panoramic:
+                logger.info("Planar or panoramic; pose from homography currently not supported.")
+            if insufficient_inliers:
+                logger.info("Insufficient number of inliers.")
+
+            i2Ri1 = None
+            i2Ui1 = None
+            v_corr_idxs = np.array([], dtype=np.uint64)
+            # remove mention of errors in the report
+            two_view_report.R_error_deg = None
+            two_view_report.U_error_deg = None
+
+
+        two_view_report.i2Ri1 = i2Ri1
+        two_view_report.i2Ui1 = i2Ui1
+
+        return i2Ri1, i2Ui1, v_corr_idxs, two_view_report
+
+
+def generate_two_view_report(
+    inlier_ratio_est_model: float,
+    R_error_deg: float,
+    U_error_deg: float,
+    number_correct: int,
+    inlier_ratio: float,
+    v_corr_idxs: np.ndarray,
+) -> TwoViewEstimationReport:
+    """ """
+    two_view_report = TwoViewEstimationReport(
```

Comment:
Do we need a helper for this?

Reply:
so the problem is that this is in the dask loop, so we have to wrap with `dask.delayed()` any function we call inside `create_computational_graph()`.

Reply:
And `TwoViewEstimationReport` cannot be the function?

Reply:
it's strange, but if i wrap the class constructor in dask.delayed, I get a nasty error:
```
distributed.worker - WARNING - Compute Failed
Function:  execute_task
args:      ((<function apply at 0x104155b80>, <class 'type'>, (), (<class 'dict'>, [['inlier_ratio_est_model', None], ['num_inliers_gt_model', None], ['inlier_ratio_gt_model', None], ['R_error_deg', None], ['U_error_deg', None], ['i2Ri1', None], ['i2Ui1', None]])))
kwargs:    {}
Exception: TypeError('type() takes 1 or 3 arguments')

Traceback (most recent call last):
  File "gtsfm/runner/run_scene_optimizer_colmaploader.py", line 102, in <module>
    run_scene_optimizer(args)
  File "gtsfm/runner/run_scene_optimizer_colmaploader.py", line 49, in run_scene_optimizer
    sfm_result = sfm_result_graph.compute()
  File "/Users/johnlambert/anaconda3/envs/gtsfm-v2/lib/python3.8/site-packages/dask/base.py", line 285, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/Users/johnlambert/anaconda3/envs/gtsfm-v2/lib/python3.8/site-packages/dask/base.py", line 567, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/Users/johnlambert/anaconda3/envs/gtsfm-v2/lib/python3.8/site-packages/distributed/client.py", line 2672, in get
    results = self.gather(packed, asynchronous=asynchronous, direct=direct)
  File "/Users/johnlambert/anaconda3/envs/gtsfm-v2/lib/python3.8/site-packages/distributed/client.py", line 1981, in gather
    return self.sync(
  File "/Users/johnlambert/anaconda3/envs/gtsfm-v2/lib/python3.8/site-packages/distributed/client.py", line 852, in sync
    return sync(
  File "/Users/johnlambert/anaconda3/envs/gtsfm-v2/lib/python3.8/site-packages/distributed/utils.py", line 354, in sync
    raise exc.with_traceback(tb)
  File "/Users/johnlambert/anaconda3/envs/gtsfm-v2/lib/python3.8/site-packages/distributed/utils.py", line 337, in f
    result[0] = yield future
  File "/Users/johnlambert/anaconda3/envs/gtsfm-v2/lib/python3.8/site-packages/tornado/gen.py", line 762, in run
    value = future.result()
  File "/Users/johnlambert/anaconda3/envs/gtsfm-v2/lib/python3.8/site-packages/distributed/client.py", line 1846, in _gather
    raise exception.with_traceback(traceback)
  File "/Users/johnlambert/anaconda3/envs/gtsfm-v2/lib/python3.8/site-packages/dask/utils.py", line 34, in apply
    return func(*args, **kwargs)
TypeError: type() takes 1 or 3 arguments
```
So I prefer to just use the helper.